src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c

```diff
@@ -0,0 +1,375 @@
+}
+
+static void* HandleFromKeys(JNIEnv* env,
+                            jobject /*PublicKey*/ publicKey,
+                            jobject /*PrivateKey*/ privateKey,
+                            PAL_KeyAlgorithm* algorithm)
+{
+    if ((*env)->IsInstanceOf(env, privateKey, g_DSAKeyClass))
+    {
+        *algorithm = PAL_DSA;
+        return AndroidCryptoNative_CreateKeyPair(env, publicKey, privateKey);
+    }
+    else if ((*env)->IsInstanceOf(env, privateKey, g_ECKeyClass))
+    {
+        *algorithm = PAL_EC;
+        return AndroidCryptoNative_NewEcKeyFromKeys(env, publicKey, privateKey);
+    }
+    else if ((*env)->IsInstanceOf(env, privateKey, g_RSAKeyClass))
+    {
+        *algorithm = PAL_RSA;
+        return AndroidCryptoNative_NewRsaFromKeys(env, publicKey, privateKey);
+    }
+
+    LOG_INFO("Ignoring unknown privake key type");
+    return NULL;
+}
+
+static int32_t EnumerateCertificates(JNIEnv* env, jobject /*KeyStore*/ store, EnumCertificatesCallback cb, void* context)
+{
+    int32_t ret = FAIL;
+
+    // Enumeration<String> aliases = store.aliases();
+    jobject aliases = (*env)->CallObjectMethod(env, store, g_KeyStoreAliases);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // while (aliases.hasMoreElements()) {
+    //     String alias = aliases.nextElement();
+    //     KeyStore.Entry entry = store.getEntry(alias);
+    //     if (entry instanceof KeyStore.PrivateKeyEntry) {
+    //         ...
+    //     } else if (entry instanceof KeyStore.TrustedCertificateEntry) {
+    //         ...
+    //     }
+    // }
+    jboolean hasNext = (*env)->CallBooleanMethod(env, aliases, g_EnumerationHasMoreElements);
+    while (hasNext)
+    {
+        INIT_LOCALS(loc, alias, entry, cert, publicKey, privateKey);
+
+        loc[alias] = (*env)->CallObjectMethod(env, aliases, g_EnumerationNextElement);
+        ON_EXCEPTION_PRINT_AND_GOTO(loop_cleanup);
+
+        loc[entry] = (*env)->CallObjectMethod(env, store, g_KeyStoreGetEntry, loc[alias], NULL);
+        ON_EXCEPTION_PRINT_AND_GOTO(loop_cleanup);
+
+        if ((*env)->IsInstanceOf(env, loc[entry], g_PrivateKeyEntryClass))
+        {
+            // Certificate cert = entry.getCertificate();
+            // Public publicKey = cert.getPublicKey();
+            // PrivateKey privateKey = entry.getPrivateKey();
+            loc[cert] = (*env)->CallObjectMethod(env, loc[entry], g_PrivateKeyEntryGetCertificate);
+            loc[publicKey] = (*env)->CallObjectMethod(env, loc[cert], g_X509CertGetPublicKey);
+            loc[privateKey] = (*env)->CallObjectMethod(env, loc[entry], g_PrivateKeyEntryGetPrivateKey);
+
+            PAL_KeyAlgorithm keyAlgorithm = PAL_UnknownAlgorithm;
+            void* keyHandle = HandleFromKeys(env, loc[publicKey], loc[privateKey], &keyAlgorithm);
+            if (keyHandle != NULL)
+            {
+                cb(AddGRef(env, loc[cert]), keyHandle, keyAlgorithm, context);
+            }
```

Comment:
This feels like it should be

```
if (keyHandle != NULL && loc[cert] != NULL)
{
   current cb
}
else if (loc[cert] != NULL)
{
   cb with cert and no key
}
```

That is

* Presumably a key-only shouldn't fire the callback with a null cert
* An unrecognized key algorithm seems like it shouldn't stop the cert from being emitted.